File: day17/day17.cpp
#include <fstream>
#include <iostream>
#include <vector>

#include <fmt/format.h>

using namespace std;

static size_t part1(size_t steps)
{
	vector<size_t> v(1);
	size_t pos = 0;
	for (size_t i = 1; i <= 2017; i++)
	{
		pos = (pos + steps) % v.size() + 1;
		v.insert(v.begin()+pos, i);
	}
	return v[(pos+1) % v.size()];
}

static size_t part2(size_t steps)
{
	size_t pos = 0;
	size_t size = 1;
	size_t last = -1;
	for (size_t i = 1; i <= 50000000ULL; i++)
	{
		pos = (pos + steps) % size + 1;
		size++;
		if (pos == 1)
		{
			last = i;
		}
	}
	return last;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	size_t steps;
	input>>steps;
	if (input.fail())
	{
		cerr << "Cannot parse the data" << endl;
		return 1;
	}
	input.close();

	cout << fmt::format("Part1: {}", part1(steps)) << endl
	     << fmt::format("Part2: {}", part2(steps)) << endl;
	return 0;
}

File: day17/day17.py
#!/usr/bin/env python3

def part1(steps):
    lst = [0]
    pos = 0
    for i in range(1, 2017+1):
        pos = (pos + steps) % len(lst) + 1
        lst.insert(pos, i)
    return lst[(pos+1)%len(lst)]

def part2(steps):
    lstlen = 1
    pos = 0
    last = None
    for i in range(1, 50000000+1):
        pos = (pos + steps) % lstlen + 1
        lstlen += 1
        if pos == 1:
            last = i
    return last

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", part1(int(txt)))
    print("Part2:", part2(int(txt)))
